### DISCUSSION TITLE
Notebook 4 (Transformers) not working

### DISCUSSION THREAD

**Discussion by robertderidder** [TUE-ARIA/NNCV#31] [Q&A] [answered]
Hello, I have trouble with notebook 4. 
I successfully completed the "Class Attention(nn.Module)" and "class PatchEmbed(nn.Module)" (or at least it passed the tests in the notebook). 
I then tried to train the ViT from scratch but got accuracies o 10.0% each time. I thought maybe this is the point of the notebook to show that training is not feasible on a laptop.  Tim suggested to change "own_model = vit_small(patch_size=16)" to  "own_model = vit_small(patch_size=16, num_classes=10)" But this doesn't work.  

Then I moved on to the pre-trained network. Just executing the code gives an initial accuracy of 0.0% and this becomes 10% when trained. So again no improvement. An instructor was confused because after loading the model, the output states:
VisionTransformer(
               skip all this
            (norm): LayerNorm((384,), eps=1e-06, elementwise_affine=True)
            (head): Identity()
            )

So the head is just an identity matrix? 

Maybe the exercise was not just to run all this code, but to add or change some elements. If that is the case, it is not clear to me. 
Hopefully you can clear it up. 
Thanks

**Comment by chjkusters** [comment, marked as answer]
We have reviewed both the notebook and the utils file and did not find any issues. However, you do need to specify num_classes when calling vit_small. As you correctly pointed out, it is possible to pass the unit tests even if the implementation is incorrect, as the tests verify output types and shapes. We recommend attending tomorrow's computer class to double-check your implementation together.

**Comment by robertderidder** [comment]
Thanks! I'll be there. Probably something to do with my implementation then
